src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs

```diff
@@ -0,0 +1,346 @@
+/// of its key. This works by having a list of partition, each is a stack of Job where each partition holds only Jobs for
+/// a portion of the address space. The partition are ordered in increasing order. On run, each worker will go through
+/// the partitions one by one in ascending order. The pointer to which partition to process next is shared with all worker
+/// so that all worker works on a small common range of the address space.
+///
+/// The working set, which is the number of Job waiting to be processed is adjusted by increasing the partition number
+/// and the max batch size (num of job to be processed on each worker visit to a partition).
+/// If the size of working set is large enough, the required data between Job in a batch is available near each other
+/// which increases cache hit. The exact minimum size of working set depends on the size of the database, which seems
+/// to increase linearly. For goerli, that seems to be 256MB, and mainnet it seems to be 1Gb. Increasing the working
+/// set decreases reads, however up to a certain point, the time taken for writes is much higher than read, so no
+/// point in increasing memory budget further. For goerli, that seems to be 3GB and for mainnet, that seems to be 8Gb.
+/// </summary>
+public class BatchedTrieVisitor
+{
+    // Not using shared pool so GC can reclaim them later.
+    private ArrayPool<Job> _jobArrayPool = ArrayPool<Job>.Create();
+    private ArrayPool<(TrieNode, SmallTrieVisitContext)> _trieNodePool = ArrayPool<(TrieNode, SmallTrieVisitContext)>.Create();
+
+    private readonly int _maxBatchSize;
+    private readonly long _partitionCount;
+    private readonly CompactStack<Job>[] _partitions;
+    private readonly long _targetCurrentItems;
+
+    private long _activeJobs;
+    private long _queuedJobs;
+    private bool _failed;
+    private long _currentPointer;
+
+    private readonly ITrieNodeResolver _resolver;
+    private readonly ITreeVisitor _visitor;
+
+    public BatchedTrieVisitor(
+        ITreeVisitor visitor,
+        ITrieNodeResolver resolver,
+        VisitingOptions visitingOptions)
+    {
+        _visitor = visitor;
+        _resolver = resolver;
+
+        _maxBatchSize = 128;
+
+        // The keccak + context itself should be 40 byte. But the measured byte seems to be 52 from GC stats POV.
+        // The * 2 is just margin. RSS is still higher though, but that could be due to more deserialization.
+        long recordSize = 52 * 2;
+        long recordCount = visitingOptions.FullScanMemoryBudget / recordSize;
+
+        // Generally, at first, we want to attempt to maximize number of partition. This tend to increase throughput
+        // compared to increasing batch size.
+        _partitionCount = recordCount / _maxBatchSize;
+
+        // 3000 is about the num of file for state on mainnet, so we assume 4000 for an unpruned db. Multiplied by
```

Comment:
so if I understand the comment you chose the best value for mainnet - 4000. However, what about other chains? Should we make it configurable per chain? Of course, mainnet >>>> all ethereum testnets. However, good support for Gnosis chain is expected.